selene-lib/src/rules/duplicate_condition.rs

```diff
@@ -0,0 +1,336 @@
+                    tokenizer::TokenType::Symbol { symbol } => match symbol {
+                        tokenizer::Symbol::False | tokenizer::Symbol::Nil => Some(qmc::Bool::False),
+                        tokenizer::Symbol::True => Some(qmc::Bool::True),
+
+                        other => unreachable!("Reached unknown token type {}", other),
+                    },
+
+                    tokenizer::TokenType::Identifier { .. } => self.grab_literal(expression),
+
+                    // Any other value we can get here is truthy
+                    _ => Some(qmc::Bool::True),
+                },
+
+                ast::Value::ParenthesesExpression(expression) => {
+                    self.expression_to_qmc_bool(expression)
+                }
+
+                ast::Value::Var(_) => self.grab_literal(expression),
+
+                _ => None,
+            },
+
+            _ => None,
+        }
+    }
+
+    fn format_bool(&self, to_format: qmc::Bool) -> Option<String> {
+        Some(
+            match to_format {
+                qmc::Bool::Term(term) => Some(
+                    self.literals
+                        .get(term as usize)
+                        .expect("received term that wasn't inside literals")
+                        .to_string(),
+                ),
+
+                qmc::Bool::And(bools) => self.format_bools_with_separator(bools, "and"),
+                qmc::Bool::Or(bools) => self.format_bools_with_separator(bools, "or"),
+
+                qmc::Bool::Not(other_bool) => {
+                    Some(format!("not {}", self.format_bool(*other_bool)?))
+                }
+
+                // We provide `false` and `true` for any falsy or truthy values, but we don't have
+                // any recollection of what those *were*.
+                qmc::Bool::False | qmc::Bool::True => None,
+            }?
+            .trim()
+            .to_owned(),
+        )
+    }
+
+    fn format_bools_with_separator(
+        &self,
+        bools: Vec<qmc::Bool>,
+        separator: &str,
+    ) -> Option<String> {
+        Some(
+            bools
+                .into_iter()
+                .rev()
+                .filter_map(|condition| self.format_bool(condition))
+                .map(|formatted| formatted.trim().to_owned())
+                .collect::<Vec<_>>()
+                .join(&format!(" {} ", separator)),
+        )
+    }
+}
+
+impl DuplicateConditionVisitor {
+    fn check_expression(&mut self, expression: &ast::Expression) {
+        let mut transformer = QmcBoolTransformer::default();
+        let qmc_bool = match transformer.expression_to_qmc_bool(expression) {
+            Some(qmc_bool) => qmc_bool,
+            None => return,
+        };
+
+        if transformer.layers == 0 {
+            return;
+        }
+
+        let simplified = qmc_bool.simplify();
+
+        if simplified.len() > 1 {
+            unimplemented!("complicated expression: {}", expression);
```

Comment:
I know absolutely nothing about QMC and am just hoping selenology will find real code that hits this.